src/charm.py

```diff
@@ -460,27 +460,76 @@ def _iscsi_initiator(self, tenv: Environment) -> None:
+        # and config file doesn't contain initiator name
+        if not initiator_name and not initiator_name_from_file:
+            # generate random iqn
             initiator_name = subprocess.getoutput("/sbin/iscsi-iname")
             logging.warning(
-                "The hostname was not found in the initiator dictionary! "
-                "The random iqn %s will be used for %s",
+                "Hostname was not found in initiator-dict and config file. "
+                + "The randomly generated iqn %s will be used for %s",
                 initiator_name,
                 hostname,
             )
-
-        logging.info("Rendering initiatorname.iscsi")
-        ctxt = {"initiator_name": initiator_name}
-        template = tenv.get_template("initiatorname.iscsi.j2")
-        rendered_content = template.render(ctxt)
-        self.ISCSI_INITIATOR_NAME.write_text(rendered_content)
+            render_file = True
+        elif initiator_name and initiator_name != initiator_name_from_file:
+            # initiator name configuration is provided and isn't the same as
+            # what's already present in the initiatorname.iscsi file.
+            # so render file with provided name
+            render_file = True
+        else:
+            # do not render the file again for these cases:
+            # 1. initiator name configuration from initiator-dictionary is same as
+            #    name in initiatorname.iscsi file
+            # 2. no initiator name configuration but name is present in file. use
+            #    the same name.
+            render_file = False
+
+        if render_file:
+            logging.info("Rendering initiatorname.iscsi")
+            ctxt = {"initiator_name": initiator_name}
+            template = tenv.get_template("initiatorname.iscsi.j2")
+            rendered_content = template.render(ctxt)
+            self.ISCSI_INITIATOR_NAME.write_text(rendered_content)
+
+    def _get_initiator_name_from_file(self, iscsi_config_file: Path | None = None) -> str | None:
+        """Parse iSCSI config file to check if "InitiatorName" is set.
+
+        iSCSI config file is /etc/iscsi/initiatorname.iscsi by default.
+
+        Returns:
+            "InitiatorName" value in config file if present.
+            None if config file isn't present or "InitiatorName" isn't set.
+        """
+        # can't set default param value to instance attribute
+        if not iscsi_config_file:
+            iscsi_config_file = self.ISCSI_INITIATOR_NAME
+        try:
+            with open(iscsi_config_file, encoding="utf-8") as file:
+                lines = file.readlines()
+        except FileNotFoundError:
+            logging.warning("%s doesn't exist", iscsi_config_file)
+            return None
+
+        initiator_name, value = None, None
+        for line in lines:
+            line = line.strip()
+            if line and not line.startswith("#"):
```

Comment:
**nitpick:** 

I personal prefer to use regex instead of if-else condition to find context in the file. Also the target string in this case is not too hard to match.